ark/utils/spatial_analysis_utils.py

```diff
@@ -266,6 +251,150 @@ def compute_close_cell_num_random(marker_nums, dist_mat, dist_lim, bootstrap_num
+            # compute the number of samples we need per bootstrap
+            # given the percentages specified in cell_type_rand
+
+            # subset the distance matrix to include only the rows/cols corresponding to the
+            # intersection between marker_pos_inds and cell_type_data[cell_type]['indices'],
+
+            # use np.choice to generate the close_rand_num_hits data from the
+            # flattened distance matrix generated from above, this corresponds to
+            # our random samples for all our bootstraps for one cell_type
+
+            # add the sum of the np.choice call above to close_num_rand[j, k, :], we can add
+            # because we compute each cell_type sum independent of each other, it works the
+            # same if we aggregated random indices together and used np.choice on that
+
+            # TODO: I'll need to talk with Erin about this, but the randomization strategy
+            # still needs clarification. Currently, the percentages I'm taking are from the
+            # number of cell_type hits per marker_counts. I'm not 100% certain that this is what
+            # she was getting at. Not using m1n and m2n definitely looks suspect here.
+            # Possibly, we'll need to use the percents specified in cell_type_rand to partition
+            # m1n * m2n instead.
+
+            for cell_type in cell_type_data:
+                # generate the dimensions of our samples array for the cell_type
+                samples_per_bootstrap = int(len(cell_type_data[cell_type]['marker_inds']) *
+                                            cell_type_data[cell_type]['percent'])
+
+                samples_dim = (samples_per_bootstrap, bootstrap_num)
+
+                # now generate the marker inds
+                marker_inds_to_choose = np.random.choice(
+                    cell_type_data[str(cell_type)]['marker_inds'],
+                    samples_per_bootstrap)
+
+                # this happens if the intersection between the marker inds and the cell type inds
+                # is the null set, in this case, we do not attempt to count the
+                # number of random hits since it would just be 0 anyways
+                if len(marker_inds_to_choose) == 0:
+                    continue
+
+                # subset the distance matrix
+                dist_mat_bin_flat = dist_mat_bin.values[marker_inds_to_choose,
+                                                        marker_inds_to_choose].flatten()
+
+                # count the number of positive random hits we get
+                count_close_rand_num_hits = np.sum(
+                    np.random.choice(dist_mat_bin_flat, samples_dim, True),
+                    axis=0
+                )
+
+                # add to the corresponding enry in close_num_rand we take
+                close_num_rand[j, k, :] += count_close_rand_num_hits
+                close_num_rand[k, j, :] += count_close_rand_num_hits
+
```

Comment:
I'm a little confused about the way you're choosing things here.  So, it looks like, to me, the `marker_inds_to_choose` randomly chosen for each `cell_type` are the same for its bootstrapping sample.  Then, that fixed subset is randomly summed over as usual.

Why are we randomly subsetting marker_inds?  Don't we just randomly sum over marker_inds?  Definitely don't mind a mini-lecture if I'm wrong!

Reply:
also, if this is a non-redundant way of getting random marker indices, why are we not generating num_bootstraps * samples_per_bootstrap number of them and only have a fixed set over all bootstraps?

Reply:
Oof, yeah you make a good point. I had meant to sample different marker_inds per bootstrap_num, but it seems like I had overlooked that aspect.

Of course, randomly sampling marker_inds may be the wrong approach in the first place, as you mention. I may need to ask Erin for clarification, but maybe @ngreenwald will know something about this. The whole randomization parameter/random sampling thing has spun my head in circles today 😵.

Reply:
Yes, the only difference between this and regular channel enrichment is that instead of randomizing over all possible indices, we randomize over the indices belonging to that specific cell population. The optimization adam made for that function (generating all of the bootstrap indices in the same call, rather than a separate nested for loop), should work here as well.

Reply:
@ackagel @ngreenwald oh ok, I think I get what you're getting at. Instead of iterating over each cell_type (ugly nested for loop), we can simply generate all the bootstrap samples at once and combine the marker_inds of all the cell_types into one list that we can use to index into dist_mat_bin_flat.

If that's the case, I have three questions:

1. How would Erin's randomization strategy play into this? Currently, I'm thinking that we randomly select the marker_inds we wish for each cell_type based on the percentages the user specifies, combine them into one list, and use that to index into dist_mat_bin. Without the extra nested for loop, it makes sense that we now only randomly choose values within the indices of the dist_mat we're interested in.
2. Is it fine to leave out m1n and m2n completely? We're not working with the total marker count per marker_num, rather with bits and pieces of them (as in, total marker count for cell_type 1, total marker count for cell_type 2, ...), and we don't need m1n and m2n to compute these values. If I'm missing something please let me know.
3. Would it be better if we allowed the user to specify lists of FlowSOM values they wanted to randomize over? For example, one category could be [1, 2], another could be [4, 5, 6], and another could be [9, 10], each with an associated percentage? That way, we can avoid having to specify an else category and we can increase the flexibility of the randomization process beyond just 1 FlowSOM ID per category. What do you think?

Reply:
What I was referring to earlier was that there was a for loop over `boostrap_num`. I think you've changed that now, which is great. That was the main inefficiency. 

It's currently a bit hard for me to figure out what's going on here. The wall of text run-on sentence comments don't actually help to explain what the code does. I think the easiest thing will be to wait until you have the example data from Erin so that we can check for correctness while we're making optimization changes